transport/src/main/java/io/netty/channel/DefaultChannelPromiseAggregatorFactory.java

```diff
@@ -0,0 +1,173 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.channel;
+
+import io.netty.util.concurrent.EventExecutor;
+
+/**
+ * Provides the ability to associate the outcome of multiple {@link ChannelPromise}
+ * objects into a single {@link ChannelPromise} object.
+ * <p>
+ * Not thread safe. Must be used within the {@link EventLoop}.
+ */
+public final class DefaultChannelPromiseAggregatorFactory extends DefaultChannelPromise
+                                                          implements ChannelPromiseAggregatorFactory {
+    private final ChannelPromise promise;
+    private int expectedCount;
+    private int doneCount;
+    private Throwable lastFailure;
+    private boolean doneAllocating;
+
+    /**
+     * Create a new instance which returns an appropriate type if {@code aggregationResult} is {@link #isVoid()}.
+     * @param aggregationResult The promise whose status will reflect the result of the aggregation.
+     * @param executor the {@link EventExecutor} which is used to notify the promise once it is complete.
+     */
+    public static ChannelPromiseAggregatorFactory newInstance(ChannelPromise aggregationResult,
+                                                              EventExecutor executor) {
+        return aggregationResult.isVoid() ? new VoidChannelPromiseAggregatorFactory(aggregationResult.channel(), true)
+                                          : new DefaultChannelPromiseAggregatorFactory(aggregationResult, executor);
+    }
+
+    /**
+     * Create a new instance.
+     * @param aggregationResult The promise whose status will reflect the result of the aggregation.
+     * @param executor the {@link EventExecutor} which is used to notify the promise once it is complete.
+     */
+    private DefaultChannelPromiseAggregatorFactory(ChannelPromise aggregationResult, EventExecutor executor) {
+        super(aggregationResult.channel(), executor);
+        assert !aggregationResult.isDone();
+        this.promise = aggregationResult;
+    }
+
+    /**
+     * Allocate a new promise which will be used to aggregate the overall success of this promise aggregator.
+     * @return A new promise which will be aggregated.
+     * {@code null} if {@link #doneAllocatingPromises()} was previously called.
+     */
+    public ChannelPromise newPromise() {
+        assert !doneAllocating : "Done allocating. No more promises can be allocated.";
+        ++expectedCount;
+        return this;
+    }
+
+    /**
+     * Signify that no more {@link #newPromise()} allocations will be made.
+     * The aggregation can not be successful until this method is called.
+     * @return The promise that is the aggregation of all promises allocated with {@link #newPromise()}.
+     */
+    public ChannelPromise doneAllocatingPromises() {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add the "@Override" annotation above this method signature [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1161)